### ISSUE TITLE
Make linkage errors from accessing Java classes more consistent.

### ISSUE THREAD

**Issue by headius** [jruby/jruby#799]
When there's a linkage error loading a Java class, Java reports an exception that indicates the class it could not find (usually because a required jar file was not in classpath). For java_import, we scrape this class name out of the message for our error.

Example:

```
ext-jruby-local ~/projects/jruby $ jar tf blah.jar
META-INF/
META-INF/MANIFEST.MF
Blah.class

ext-jruby-local ~/projects/jruby $ jar tf foo.jar
META-INF/
META-INF/MANIFEST.MF
Foo.class

ext-jruby-local ~/projects/jruby $ cat Blah.java
public class Blah {
  static {
    Foo.foo();
  }
}

ext-jruby-local ~/projects/jruby $ cat Foo.java
public class Foo {
  public static void foo() {
    System.out.println("foo");
  }
}
```

If we java_import the class by name, the error reports that it's probably the Foo class missing:

```
ext-jruby-local ~/projects/jruby $ jruby -e "require 'blah.jar'; java_import 'Blah'"
NameError: cannot link Java class Blah, probable missing dependency: Foo
         for_name at org/jruby/javasupport/JavaClass.java:1242
  get_proxy_class at org/jruby/javasupport/JavaUtilities.java:34
      java_import at file:/Users/headius/projects/jruby/lib/jruby.jar!/jruby/java/core_ext/object.rb:26
              map at org/jruby/RubyArray.java:2412
      java_import at file:/Users/headius/projects/jruby/lib/jruby.jar!/jruby/java/core_ext/object.rb:22
           (root) at -e:1
```

However, if we access it using the Java namespace (as a lazy constant), we do not include this information:

```
ext-jruby-local ~/projects/jruby $ jruby -e "require 'blah.jar'; java_import Java::Blah"
NameError: cannot link Java class Blah
  get_top_level_proxy_or_package at org/jruby/javasupport/JavaUtilities.java:49
                   const_missing at file:/Users/headius/projects/jruby/lib/jruby.jar!/jruby/java/java_module.rb:4
                          (root) at -e:1
```

The error logic for both paths should be unified so we have consistent errors.

**Comment by headius** [comment]
Changes over the years have improved this in some ways and degraded it in others.

The second form now includes information about the `NoClassDefFound` preventing `Blah` from linking:

```
$ jruby -e 'Java::Blah'
NameError: cannot link Java class Blah (java.lang.NoClassDefFoundError: Foo)
  const_missing at org/jruby/javasupport/Java.java:1298
         <main> at -e:1
```

As does the package syntax that confusingly seemed to be unable to find `Blah` in a package from #1010:

```
$ jruby -e 'Java::blah.Blah'
NameError: cannot link Java class blah.Blah (java.lang.NoClassDefFoundError: blah/Foo)
  method_missing at org/jruby/javasupport/JavaPackage.java:253
          <main> at -e:1
```

But surprisingly, around the release of 9.3 we lost the error for the `java_import` forms:

```
$ jruby -e 'java_import "Blah"'
$ jruby -e 'java_import "blah.Blah"'
$
```

I believe the regression was probably a small mistake porting the Ruby version of `java_import` to Java in #6661 here, where it loads but does not initialize the class (so it does not link):

https://github.com/jruby/jruby/commit/5a8ef9fea5333abae43c9917676b1df75e7aa22a#diff-cabfd925021b0b615287200ad6cf2469755af74bf4dfd560ea31161b22fda59eR116

I'm going to call this one fixed as of 9.3.0.0 because the errors are consistent, but open an issue for the regression. The errors are now consistent, but due to the non-initialization during `java_import` the error for the first form is deferred until first use:

```
$ (chruby jruby-9.3.14.0 ; jruby -e 'java_import "blah.Blah"; Blah.new')
Unhandled Java exception: java.lang.NoClassDefFoundError: blah/Foo
java.lang.NoClassDefFoundError: blah/Foo
               <clinit> at blah/Blah.java:5
...
```

**Comment by headius** [comment]
The regression described in https://github.com/jruby/jruby/issues/8299 has been fixed for JRuby 10.
